### PR TITLE
Intercept template config

### PR DESCRIPTION
**Pull request by ProxiBlue**
The current iterable extension is limited in which templates can be sent to campaigns.

This merge propose improvement to allow any magento transactional email to be intercepted, and diverted to an iterable campaign, simply by using a template => campaign id configurable map in admin.

If a transactional email  (designated by template name) exists in the map, and has a valid campaign id, it will be diverted to the given campaign, and not sent as a normal transactional email.
The transactional magento email is used as a fallback, in case there is an issue.

This requires the rewrite for core email sending of class  Mage_Core_Model_Email_Template to be enabled 

``` xml
<core>
    <rewrite>
       <email_template>Iterable_TrackOrderPlaced_Model_Email_Template</email_template>
    </rewrite>
</core>
```

Care must be taken to ensure that no other extension exists that also rewrites that class.
An example of this can be 'Aschroder_SMTPPro'

If there is a rewrite conflict, it is vital that the conflict is addressed in the manner that iterable extension rewrites the core class, and have the iterable class to extend the other extension template class
This is to ensure that the iterable extension first determines if the email must be diverted, before passing code flow to the other (or core) class

An example to resolve the conflict with  Aschroder_SMTPPro:
- Edit /app/code/community/Aschroder/SMTPPro/etc/config.xml and remove rewrite directive

``` xml
<core>
   <rewrite>
    <email>Aschroder_SMTPPro_Model_Email</email>
        <mail_template>Aschroder_SMTPPro_Model_Email_Template</email_template>
   </rewrite>
</core>
```

will become 

``` xml
<core>
   <rewrite>
    <email>Aschroder_SMTPPro_Model_Email</email>
        <!-- CONFLICT RESOLUTION -->
            <!-- Iterable extension needs to rewrite this core class first ->
        <!--<email_template>Aschroder_SMTPPro_Model_Email_Template</email_template>-->
  </rewrite>
</core>
```

Next make the iterable extension class   `Iterable_TrackOrderPlaced_Model_Email_Template` extend the class `Aschroder_SMTPPro_Model_Email_Template`

Now, the core class is rewritten by iterable, dealt with by iterable logic, and if need be flow is passed to the Aschroder_SMTPPro (or core) extension.

### CODE REVIEW

File: app/code/community/Iterable/TrackOrderPlaced/Helper/Data.php [hunk 1]
```diff
@@ -1,106 +1,69 @@
...
     }
 
-    private function getIterableApiToken() {
+    private function getIterableApiToken()
+    {
         $apiKeyJson = $this->getDecodedMagentoApiToken();
-        if ($apiKeyJson == NULL) {
-            return NULL;
+        if ($apiKeyJson == null) {
+            mage::throwException('API Key is not set');
```
Comment (dontgitit) [inline app/code/community/Iterable/TrackOrderPlaced/Helper/Data.php]: I'm rather hesitant to throw an exception here if there's no API key configured; there's code in callIterableApi that checks for `null` right after this anyway. Additionally, if they don't take their store offline while installing the extension, then I think users are going to see exceptions until an API key is entered?

Comment (ProxiBlue) [inline app/code/community/Iterable/TrackOrderPlaced/Helper/Data.php]: Valid point. I will adjust. Mote that i was going to withdraw rhis merge in favour of a new one with improved code and ability to push events as well.

Comment (dontgitit) [inline app/code/community/Iterable/TrackOrderPlaced/Helper/Data.php]: Sounds good, I'll wait for that. The rest of this looks awesome! :)

**Comment by ProxiBlue** [comment]
I have updated the PR with new functionality: Ability to map email templates (sending) to events.

**Comment by dontgitit** [comment]
LGTM. Will test locally and release new version.

**Comment by dontgitit** [comment]
Hmm, what was the intention behind the optional campaign and template id for event rewriting? Those are used for campaign attribution; I don't think it ever makes sense to set them to static values.

On a sidenote, we set several cookies on click, and those should probably be the campaignId/templateId set in the `track()` call, similar to what's done in `Observer->orderPlaced`:

```
        $cookieModel = Mage::getModel('core/cookie');
        $campaignId = $cookieModel->get('iterableEmailCampaignId');
        $campaignId = empty($campaignId) ? NULL: intval($campaignId);
        $templateId = $cookieModel->get('iterableTemplateId');
        $templateId = empty($templateId) ? NULL: intval($templateId);
```

I'm going to switch `Helper->track` to do that as well (unless there was a good reason you wanted static values for those). 

Looks good so far otherwise!

**Comment by ProxiBlue** [comment]
The API docs located here: http://support.iterable.com/hc/en-us/articles/204780579-API-Overview-and-Sample-Payloads#events shows passing of those as part of the event data.

There was no explanation about those (or the userid), so i figured I'd just add ability to set those and pas them as per the api example.

**Comment by dontgitit** [comment]
Got it! Yeah, they should be auto-set from the cookies. I'll fix that, sorry.

**Comment by ProxiBlue** [comment]
No worries

I have a small addition, which I will PR later. It is basically ability to push array of data into the registry (lets say from a custom controller) - as the event send, this will be merged with the pushed data.

This will give users ability to push in additional data before event is fired

**Comment by dontgitit** [comment]
I think there were also cases where `data` wasn't set in the variables, and it was crapping out in sendAsEvent; fixing it by instantiating `$data` to an empty array.

**Comment by dontgitit** [comment]
Made some fixes/changes here: https://github.com/Iterable/magento-extension/commit/8dee7edf871b6ca84842419c3b8cc090c8856a94

**Pull request merged**